src/wg/util/allocator.d

```diff
@@ -0,0 +1,70 @@
+module wg.util.allocator;
+
+import wg.util.traits : isClassOrStructPtr;
+
+/**
+ * A 'hard' allocator struct that supports `std.experimental.allocator`, and can easily interface with C
+ */
+struct Allocator
+{
+    alias AllocFunc =  void* function(size_t bytes, size_t* allocated, shared void* userData) @nogc nothrow;
+    alias FreeFunc =  void function(void* alloc, size_t bytes, shared void* userData) @nogc nothrow;
+
+    AllocFunc allocFunc;
+    FreeFunc freeFunc;
+    shared void* userData;
+
+    void[] allocate(size_t bytes) @nogc nothrow
+    {
+        void* mem = allocFunc(bytes, &bytes, userData);
+        return mem[0 .. bytes];
+    }
+
+    bool deallocate(void[] mem) @nogc nothrow
+    {
+        freeFunc(mem.ptr, mem.length, userData);
+        return true;
+    }
+}
+
+/**
+ * make an `Allocator` from an object that conforms to the `std.experimental.allocator` interface
+ */
+Allocator makeAllocator(Alloc)(shared(Alloc) instance = null) if (isClassOrStructPtr!Alloc)
+{
+    static alloc(size_t bytes, size_t* allocated, shared void* userData) @nogc nothrow
+    {
+        shared Alloc allocator = cast(shared Alloc)userData;
```

Comment:
What if `userData` is null? (and `instance` is null by default)

Reply:
Then `Alloc` is null...? What do you mean?

Reply:
I guess he like me doesn't understand why we need that null default value for `instance` since the code seem not to work when the that value is null.

Reply:
I think null instance is probably more common. Any allocator that doesn't have state... Malloc, gcalloc, mmap, etc. They have no state.

Reply:
Ah, of course. I just never thought of calling a static method through a variable that I know can be null so it took me a bit to wrap my head around it :)

Reply:
i mean, i could put a sea of `if`'s and `static if`'s, but i don't really see the advantage of that noise.

Reply:
`enforce(userData, "Allocator may not be null");` would be good enough. These details make experience much smother but probably they can be touched later.

Reply:
It may be `null`, that's perfectly valid.

Reply:
Ah, things are more complex than I thought, I understand you remark about a sea of `if`s.

Reply:
they're not necessary though. this works just fine.
a runtime null pointer exception is trivial to diagnose and debug. i don't think a pile of code complexity is worth it.

Reply:
I totally agree, this is a minor detail and can be enhanced later.